File: go-client-sdk/pkg/models/shared/subscription.go
package shared

import (
	"time"
)

type SubscriptionFixedFeeQuantitySchedule struct {
	EndDate   *time.Time `json:"end_date,omitempty"`
	PriceID   *string    `json:"price_id,omitempty"`
	Quantity  *float64   `json:"quantity,omitempty"`
	StartDate *time.Time `json:"start_date,omitempty"`
}

type SubscriptionRedeemedCoupon struct {
	CouponID  *string    `json:"coupon_id,omitempty"`
	EndDate   *time.Time `json:"end_date,omitempty"`
	StartDate *time.Time `json:"start_date,omitempty"`
}

type SubscriptionStatusEnum string

const (
	SubscriptionStatusEnumActive   SubscriptionStatusEnum = "active"
	SubscriptionStatusEnumEnded    SubscriptionStatusEnum = "ended"
	SubscriptionStatusEnumUpcoming SubscriptionStatusEnum = "upcoming"
)

// Subscription
// A subscription represents the purchase of a plan by a customer.
//
// By default, subscriptions begin on the day that they're created and renew automatically for each billing cycle at the cadence that's configured in the plan definition.
//
// Subscriptions also default to **beginning of month alignment**, which means the first invoice issued for the subscription will have pro-rated charges between the `start_date` and the first of the following month. Subsequent billing periods will always start and end on a month boundary (e.g. subsequent month starts for monthly billing).
//
// Depending on the plan configuration, any _flat_ recurring fees will be billed either at the beginning (in-advance) or end (in-arrears) of each billing cycle. Plans default to **in-advance billing**. Usage-based fees are billed in arrears as usage is accumulated. In the normal course of events, you can expect an invoice to contain usage-based charges for the previous period, and a recurring fee for the following period.
type Subscription struct {
	ActivePlanPhaseOrder          *float64                               `json:"active_plan_phase_order,omitempty"`
	AutoCollection                *bool                                  `json:"auto_collection,omitempty"`
	CreatedAt                     time.Time                              `json:"created_at"`
	CurrentBillingPeriodEndDate   *time.Time                             `json:"current_billing_period_end_date,omitempty"`
	CurrentBillingPeriodStartDate *time.Time                             `json:"current_billing_period_start_date,omitempty"`
	Customer                      Customer                               `json:"customer"`
	DefaultInvoiceMemo            *string                                `json:"default_invoice_memo,omitempty"`
	EndDate                       time.Time                              `json:"end_date"`
	FixedFeeQuantitySchedule      []SubscriptionFixedFeeQuantitySchedule `json:"fixed_fee_quantity_schedule"`
	ID                            string                                 `json:"id"`
	NetTerms                      *int64                                 `json:"net_terms,omitempty"`
	Plan                          Plan                                   `json:"plan"`
	RedeemedCoupon                *SubscriptionRedeemedCoupon            `json:"redeemed_coupon,omitempty"`
	StartDate                     time.Time                              `json:"start_date"`
	Status                        SubscriptionStatusEnum                 `json:"status"`
}

File: python-client-sdk/src/orbapi/models/shared/subscription.py
from __future__ import annotations
import dataclasses
import dateutil.parser
from ..shared import customer as shared_customer
from ..shared import plan as shared_plan
from dataclasses_json import Undefined, dataclass_json
from datetime import datetime
from enum import Enum
from marshmallow import fields
from orbapi import utils
from typing import Optional


@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class SubscriptionFixedFeeQuantitySchedule:
    end_date: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('end_date'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    price_id: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('price_id'), 'exclude': lambda f: f is None }})
    quantity: Optional[float] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('quantity'), 'exclude': lambda f: f is None }})
    start_date: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('start_date'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    

@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class SubscriptionRedeemedCoupon:
    coupon_id: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('coupon_id'), 'exclude': lambda f: f is None }})
    end_date: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('end_date'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    start_date: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('start_date'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    
class SubscriptionStatusEnum(str, Enum):
    ACTIVE = "active"
    ENDED = "ended"
    UPCOMING = "upcoming"


@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class Subscription:
    r"""Subscription
    A subscription represents the purchase of a plan by a customer.
    
    By default, subscriptions begin on the day that they're created and renew automatically for each billing cycle at the cadence that's configured in the plan definition.
    
    Subscriptions also default to **beginning of month alignment**, which means the first invoice issued for the subscription will have pro-rated charges between the `start_date` and the first of the following month. Subsequent billing periods will always start and end on a month boundary (e.g. subsequent month starts for monthly billing).
    
    Depending on the plan configuration, any _flat_ recurring fees will be billed either at the beginning (in-advance) or end (in-arrears) of each billing cycle. Plans default to **in-advance billing**. Usage-based fees are billed in arrears as usage is accumulated. In the normal course of events, you can expect an invoice to contain usage-based charges for the previous period, and a recurring fee for the following period.
    """
    
    created_at: datetime = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('created_at'), 'encoder': utils.datetimeisoformat(False), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    customer: shared_customer.Customer = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('customer') }})
    end_date: datetime = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('end_date'), 'encoder': utils.datetimeisoformat(False), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    fixed_fee_quantity_schedule: list[SubscriptionFixedFeeQuantitySchedule] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('fixed_fee_quantity_schedule') }})
    id: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('id') }})
    plan: shared_plan.Plan = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('plan') }})
    start_date: datetime = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('start_date'), 'encoder': utils.datetimeisoformat(False), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    status: SubscriptionStatusEnum = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('status') }})
    active_plan_phase_order: Optional[float] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('active_plan_phase_order'), 'exclude': lambda f: f is None }})
    auto_collection: Optional[bool] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('auto_collection'), 'exclude': lambda f: f is None }})
    current_billing_period_end_date: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('current_billing_period_end_date'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    current_billing_period_start_date: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('current_billing_period_start_date'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    default_invoice_memo: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('default_invoice_memo'), 'exclude': lambda f: f is None }})
    net_terms: Optional[int] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('net_terms'), 'exclude': lambda f: f is None }})
    redeemed_coupon: Optional[SubscriptionRedeemedCoupon] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('redeemed_coupon'), 'exclude': lambda f: f is None }})
    